### PR TITLE
Introduce redis backend for config store.

### PR DESCRIPTION
**Pull request by jmuk**
Note: this implements the exact SPI currently we have. With
some redis features, some things (like listening updates)
can be done more efficiently, but that needs to change the SPI.

I'll address that after landing this.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/694)

### CODE REVIEW

File: pkg/config/redisstore.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	}
+	if len(u.Path) > 1 {
+		dbNum, err := strconv.ParseInt(u.Path[1:], 10, 0)
+		if err != nil {
+			return fmt.Errorf("failed to parse dbNum \"%s\", it should be an integer", u.Path[1:])
+		}
+		if dbNum < 0 {
+			return fmt.Errorf("dbNum \"%s\" should be a positive integer", u.Path[1:])
+		}
+		client.Cmd("SELECT", dbNum)
```
Comment (mandarjog) [inline pkg/config/redisstore.go]: issue a ping before "selecting" a database?

Comment (mandarjog) [inline pkg/config/redisstore.go]: Move all local validation  (extracting dbNum) before reaching out to the server.

File: pkg/config/redisstore_test.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+			t.Errorf("dbnum is not an integer: %v", err)
+		} else if num != tt.dbNum {
+			t.Errorf("expected to connect %d but connected to %d", tt.dbNum, num)
+		}
+		rs.Close()
+	}
+}
+
+func TestInvalidDBNum(t *testing.T) {
+	s, err := miniredis.Run()
```
Comment (mandarjog) [inline pkg/config/redisstore_test.go]: You don't need to start miniredis to test invalid db numbers.

File: pkg/config/store_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"errors"
+	"strings"
+	"testing"
+
+	"github.com/alicebob/miniredis"
+)
+
+func TestNewStore(t *testing.T) {
```
Comment (mandarjog) [inline pkg/config/store_test.go]: If this is redis_store test, update the filename.

**Comment by mandarjog** [comment]
https://github.com/istio/mixer/issues/695

**Comment by jmuk** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/config/redisstore.go, line 47 at r1](https://reviewable.io:443/reviews/istio/mixer/694#-KjtHV4lx6HBjK6AvIsx-r1-47:-KjtrulljzExYmnWs0KF:b-896fix) ([raw file](https://github.com/istio/mixer/blob/4ab613a10c0268dd66cbc4088493bb8876ef9369/pkg/config/redisstore.go#L47)):*
<details><summary><i>Previously, mandarjog wrote…</i></summary><blockquote>

Move all local validation  (extracting dbNum) before reaching out to the server.
</blockquote></details>

Done.

---

*[pkg/config/redisstore.go, line 126 at r1](https://reviewable.io:443/reviews/istio/mixer/694#-KjtHV4lx6HBjK6AvIsx-r1-126:-Kjtrz2qtyWSPsM0VbzO:b-896fix) ([raw file](https://github.com/istio/mixer/blob/4ab613a10c0268dd66cbc4088493bb8876ef9369/pkg/config/redisstore.go#L126)):*
<details><summary><i>Previously, mandarjog wrote…</i></summary><blockquote>

best to keep a track of the number of keys we saw last time.
That way you can do
` keys := make([]string, rs.length) `  and minimize reallocating the slice.
</blockquote></details>

Done.

---

*[pkg/config/redisstore_test.go, line 123 at r1](https://reviewable.io:443/reviews/istio/mixer/694#-KjtHV4lx6HBjK6AvIsy-r1-123:-Kjts-LZ9545x1iMVoH6:b-896fix) ([raw file](https://github.com/istio/mixer/blob/4ab613a10c0268dd66cbc4088493bb8876ef9369/pkg/config/redisstore_test.go#L123)):*
<details><summary><i>Previously, mandarjog wrote…</i></summary><blockquote>

You don't need to start miniredis to test invalid db numbers.
</blockquote></details>

Done.

---

*[pkg/config/store_test.go, line 24 at r1](https://reviewable.io:443/reviews/istio/mixer/694#-KjtHV4miHAGzTUfO1g1-r1-24:-Kjts04b08JR-XbI5LeY:b-bc00et) ([raw file](https://github.com/istio/mixer/blob/4ab613a10c0268dd66cbc4088493bb8876ef9369/pkg/config/store_test.go#L24)):*
<details><summary><i>Previously, mandarjog wrote…</i></summary><blockquote>

If this is redis_store test, update the filename.
</blockquote></details>

This was actually moved from fsstore_test.go, testing NewStore function in store.go, so I think this file name is better.
And miniredis was added because otherwise NewStore() will fail.

Removed miniredis code, and modified the expected error message is like connection failure. cf. https://github.com/istio/mixer/blob/master/pkg/config/fsstore_test.go#L116

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/istio/mixer/694)*

**Comment by jmuk** [comment]
mixs server ran with a local redis DB (i.e. the process starts) but I haven't done any further yet. will check.

File: pkg/config/redisstore.go [hunk 4]
```diff
@@ -89,6 +106,7 @@ func (rs *redisStore) Set(key, value string) (index int, err error) {
 }
 
 func (rs *redisStore) List(key string, recurse bool) (keys []string, index int, err error) {
+	keys = make([]string, rs.listLength)
```
Comment (mandarjog) [inline pkg/config/redisstore.go]: should be `make([]string, 0, rs.listLength)` ?
Otherwise keys will have first 5 elements as empty right?

Comment (mandarjog) [inline pkg/config/redisstore.go]: There should be a test in the common store tests to check for this.

Comment (jmuk) [inline pkg/config/redisstore.go]: My bad, fixed. Wrote a new pattern of calling List twice.

**Comment by jmuk** [comment]
I also checked the behavior of mixs binary, and noticed that it won't work well without the default configurations. Possibly nicer to create a data importer of the config for the initial setup.

**Comment by jmuk** [comment]
Hi,

Is there anything I'm missing?

**Pull request merged**